subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/TaskExecutionSnapshot.java

```diff
@@ -20,73 +20,44 @@
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.ImmutableSortedMap;
 import com.google.common.collect.ImmutableSortedSet;
+import org.gradle.api.NonNullApi;
 import org.gradle.internal.id.UniqueId;
 
 /**
  * Immutable snapshot of the state of a task when it was executed.
  */
-public class TaskExecutionSnapshot {
-    private final UniqueId buildInvocationId;
+@NonNullApi
+public class TaskExecutionSnapshot extends AbstractTaskExecution {
```

Comment:
It would be better to not mix `TaskExecution`s with `TaskExecutionSnapshot`s.

Reply:
Note that the `TaskExecutionSnapshot` is an immutable version of `TaskExecution`. We can also only have one `TaskExecution`. This one would need to be mutable since we update it during task execution.
I would prefer storing and loading only an immutable `TaskExecution` to/from `TaskHistory`. However, we could choose a better name for `TaskExecutionSnapshot`, e.g. `PreviousTaskExecution`. WDYT?